### ISSUE TITLE
Performer: `jnp.max` used in `nonnegative_softmax_kernel_feature_creator`

### ISSUE THREAD

**Issue by wcshin-git** [google-research/google-research#753]
Hi! Thanks for the amazing works!
I have one question in  `nonnegative_softmax_kernel_feature_creator`.
I don't understand why `jnp.max` terms are used in `nonnegative_softmax_kernel_feature_creator` as below. I cannot find the corresponding equations in the original paper. Is it just for normalization? Is there a mathematical background? 
Thank you!

https://github.com/google-research/google-research/blob/c249ee982c9ca3bb0cca4788758435c87c71fc7d/performer/fast_attention/jax/fast_attention.py#L101-L109

**Comment by jw9730** [comment]
I think it is for numerical stability, as query and key features can overflow with exp().

Line 102 (applied for queries) is equivalent to dividing each element of normalizer (Q'((K')T1)) and query (Q') with a same scalar.
Line 106 (applied for keys, which will be normalized) is an often-used trick for stable normalized exp (see [this](https://stackoverflow.com/questions/23665599/numerically-stable-implementation)).

**Comment by wcshin-git** [comment]
Hi Jinwoo! Thanks for the clear explanations!
Applied to Q: Line 102 makes Q', but when calculating attention later, Q' is also in the denominator, so it doesn't affect the result.
BTW, I think line 106 also not affecting the results. We are using the maximum value of all elements(specifically, along `axis=last_dims_t + attention_dims_t`) in the `data_dash` tensor.  And K' is also in the denominator(Q'((K')T1)) when calculating attention later. Am I mistaken?
